### PR TITLE
Add documentation for I/O behavior

### PR DESCRIPTION
**Pull request by kpamnany**
# Description

Also some minor formatting changes and move `ArtifactWrappers` from `Utilities` to `InputOutput` and move `Callbacks` from `Utilities` to `Driver`.

Additionally, replace `ClimateMachine.cli()` with `ClimateMachine.init(parse_clargs = true)`.

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [X] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [X] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [X] There are no open pull requests for this already
- [X] CLIMA developers with relevant expertise have been assigned to review this submission
- [X] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Driver/Driver.jl [hunk 1]
```diff
@@ -364,12 +330,40 @@ the process, otherwise the defaulting to `ClimateMachine.Settings`.
...
         enable integration_testing
+
+Returns `nothing`, or if `parse_clargs = true`, returns parsed command line
+arguments.
 """
-function init(; init_driver::Bool = true, kwargs...)
-    # init global setting values
+function init(;
+    parse_clargs::Bool = false,
+    custom_settings::Union{Nothing, ArgParseSettings} = nothing,
```
Comment (jakebolewski) [inline src/Driver/Driver.jl:339]: should this be `custom_clargs`?

File: docs/src/APIs/Numerics/DGMethods/Courant.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Discontinuous Galerkin Methods
+
+```@meta
+CurrentModule = ClimateMachine
+```
+
+## Functions
+
+```@docs
+Courant.advective_courant
```
Comment (charleskawczynski) [inline docs/src/APIs/Numerics/DGMethods/Courant.md:10]:
```suggestion
Courant
Courant.advective_courant
```
The complaint is about the docstring for the module, which is added with simply `ClimateMachine.Courant`.

**Pull request merged**